src/librustc_typeck/astconv.rs

```diff
@@ -222,15 +223,32 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
         assert_eq!(decl_generics.has_self, self_ty.is_some());
 
         // Check the number of type parameters supplied by the user.
-        let ty_param_defs = &decl_generics.types[self_ty.is_some() as usize..];
-        if !infer_types || num_types_provided > ty_param_defs.len() {
-            check_type_argument_count(tcx, span, num_types_provided, ty_param_defs);
+        let type_params_offset = self_ty.is_some() as usize;
+        let ty_param_defs = param_counts.types - type_params_offset;
```

Comment:
Again I think using `param_counts` obscures the meaning that would be more obvious with a single loop computing 3 values.

Reply:
I just had an idea: using an inclusive `required..=accepted` range for the number of type parameters - that way, you can even use `.contains()` to check for validity of a given length (although you usually need to compare each side anyway, for user errors) - but generally, it seems like it would indicate better what's going on.

Reply:
This doesn't end up working very well, because you just end up dealing with the endpoints, and you can't mutate the endpoints of `RangeInclusive` any more.

Reply:
Ouch. What about a struct then, with two fields? Only type-checking needs it, in a few places AFAICT.